File: src/main/java/org/simbrain/world/odorworld/OdorWorldPanel.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.world.odorworld;

import kotlinx.coroutines.Dispatchers;
import org.piccolo2d.PCamera;
import org.piccolo2d.PCanvas;
import org.piccolo2d.PNode;
import org.piccolo2d.event.PInputEventListener;
import org.piccolo2d.nodes.PImage;
import org.piccolo2d.nodes.PPath;
import org.piccolo2d.util.PBounds;
import org.simbrain.network.gui.nodes.NodeHandle;
import org.simbrain.util.StandardDialog;
import org.simbrain.util.piccolo.SceneGraphBrowser;
import org.simbrain.util.piccolo.Tile;
import org.simbrain.workspace.gui.CouplingMenu;
import org.simbrain.world.odorworld.actions.*;
import org.simbrain.world.odorworld.entities.OdorWorldEntity;
import org.simbrain.world.odorworld.gui.*;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;
import java.util.List;
import java.util.Timer;
import java.util.*;
import java.util.stream.Collectors;

/**
 * <b>OdorWorldPanel</b> represent the OdorWorld.
 */
public class OdorWorldPanel extends JPanel {

    /**
     * The Piccolo PCanvas.
     */
    private final OdorWorldCanvas canvas;

    /**
     * Reference to WorkspaceComponent.
     */
    private final OdorWorldComponent component;

    /**
     * Reference to model world.
     */
    private final OdorWorld world;

    /**
     * Selection model.
     */
    private final WorldSelectionModel selectionModel;

    /**
     * Provisional interface for selecting tiles.
     */
    private PNode tileSelectionBox = null;
    private Rectangle tileSelectionModel = null;

    /**
     * Timer to update entity animations.
     */
    private Timer animationTimer;

    /**
     * Timer used for updating entity position when they are manually moved. Allows timing of
     * manualy movement to be consistent whether the workspace is running or not.
     */
    private Timer movementTimer;

    /**
     * Used for a mask that allows multiple movements to be applied at once.
     * Lower four bits are used for U,D,L,R.
     */
    private byte manualMovementKeyState;

    /**
     * List corresponding to the layers of a tmx file.
     */
    private List<PImage> layerImageList;

    void debugToolTips() {

        if (tileSelectionModel != null) {
            if (!tileSelectionModel.contains(world.getLastClickedPosition())) {
                canvas.getLayer().removeChild(tileSelectionBox);
                tileSelectionModel = null;
                tileSelectionBox = null;
            }
        }

        if (tileSelectionBox == null) {
            int tileCoordinateX =
                    (int) (world.getLastClickedPosition().getX() / world.getTileMap().getTileWidth());
            int tileCoordinateY =
                    (int) (world.getLastClickedPosition().getY() / world.getTileMap().getTileHeight());
            tileSelectionModel = new Rectangle(
                    tileCoordinateX * world.getTileMap().getTileWidth(),
                    tileCoordinateY * world.getTileMap().getTileHeight(),
                    world.getTileMap().getTileWidth(),
                    world.getTileMap().getTileHeight()
            );
            tileSelectionBox = PPath.createRectangle(
                    tileSelectionModel.getX(),
                    tileSelectionModel.getY(),
                    tileSelectionModel.getWidth(),
                    tileSelectionModel.getHeight()
            );
            ((PPath) tileSelectionBox).setStrokePaint(Color.ORANGE);
            tileSelectionBox.setPaint(null);
            canvas.getLayer().addChild(tileSelectionBox);
        }    }

    /**
     * Extend PCanvas for custom handling of tooltips
     */
    private class OdorWorldCanvas extends PCanvas {
        @Override
        public String getToolTipText(MouseEvent event) {

            List<Tile> selectedTiles = getTileStack(event.getPoint());
            if (selectedTiles == null) {
                return "";
            }
            StringBuilder sb = new StringBuilder("Tile Ids: ");
            selectedTiles.stream().filter(Objects::nonNull).forEach(tile -> sb.append(" (" + tile.getId() + ")"));

            return sb.toString();
        }

    }

    /**
     * Construct a world, set its background color.
     *
     * @param world the frame in which this world is rendered
     */
    public OdorWorldPanel(OdorWorldComponent component, OdorWorld world) {

        this.component = component;
        this.world = world;

        canvas = new OdorWorldCanvas();
        ToolTipManager.sharedInstance().registerComponent(canvas);

        setLayout(new BorderLayout());
        this.add("Center", canvas);

        canvas.setFocusable(true);

        // Add tile map
        layerImageList = world.getTileMap().createImageList();
        canvas.getLayer().addChildren(layerImageList);

        // Remove default event handlers
        PInputEventListener panEventHandler = canvas.getPanEventHandler();
        PInputEventListener zoomEventHandler = canvas.getZoomEventHandler();
        canvas.removeInputEventListener(panEventHandler);
        canvas.removeInputEventListener(zoomEventHandler);

        // TODO: Currently disabling zoom
        // because zooming is acting wonky, esp. when you select an entity.
        // When reimplement also disallow zooming that makes the world smaller than the parent panel

        // PMouseWheelZoomEventHandler zoomHandler = new PMouseWheelZoomEventHandler();
        // zoomHandler.zoomAboutMouse();
        // canvas.addInputEventListener(zoomHandler);

        selectionModel = new WorldSelectionModel(this);
        selectionModel.addSelectionListener((e) -> {
            updateNodeHandles(e);
        });

        // Add key bindings
        KeyBindings.addBindings(this);

        // Mouse events
        canvas.addInputEventListener(new WorldMouseHandler(this, world));
        canvas.addInputEventListener(new WorldContextMenuEventHandler(this, world));

        // PCamera camera = canvas.getCamera();

        world.getEvents().getEntityAdded().on(Dispatchers.getMain(), e -> {
            EntityNode node = new EntityNode(world, e);
            canvas.getLayer().addChild(node);
            selectionModel.setSelection(Collections.singleton(node)); // not working
            repaint();
        });
        world.getEvents().getEntityRemoved().on(Dispatchers.getMain(), e -> {
            var entityNode = canvas.getLayer().getAllNodes()
                    .stream().filter(n -> n instanceof EntityNode)
                    .filter(n -> ((EntityNode)n).getEntity() == e)
                    .findFirst();
            if (entityNode.isPresent()) {
                canvas.getLayer().removeChild((PNode) entityNode.get());
                repaint();
            }
        });
        world.getEvents().getUpdated().on(Dispatchers.getMain(), true, this::centerCameraToSelectedEntity);
        world.getEvents().getFrameAdvanced().on(Dispatchers.getMain(), () -> {
            canvas.getLayer().getChildrenReference().stream()
                    .filter(i -> i instanceof EntityNode)
                    .forEach(i -> ((EntityNode) i).advance());
            repaint();
        });
        world.getEvents().getAnimationStopped().on(Dispatchers.getMain(), () -> {
            // When movement is stopped use the "static" animation so we don't show entities in strange
            // intermediate states
            canvas.getLayer().getChildrenReference().stream()
                    .filter(i -> i instanceof EntityNode)
                    .forEach(i -> ((EntityNode) i).resetToStaticFrame());
            repaint();
        });

        // Full tile map update
        world.getEvents().getTileMapChanged().on(Dispatchers.getMain(), () -> {
            renderAllLayers(world);
        });

        world.getTileMap().getEvents().getLayerAdded().on(Dispatchers.getMain(), () -> {
            renderAllLayers(world);
        });

        world.getTileMap().getEvents().getMapSizeChanged().on(Dispatchers.getMain(), () -> {
            world.events.getTileMapChanged().fireAndBlock();
        });

        // Single layer update
        world.getTileMap().getEvents().getLayerImageChanged().on(Dispatchers.getMain(), (oldImage, newImage) -> {
                int index = canvas.getLayer().indexOfChild(oldImage);
                canvas.getLayer().removeChild(oldImage);
                if (index != -1) {
                    canvas.getLayer().addChild(index, newImage);
                } else {
                    canvas.getLayer().addChild(newImage);
                }
        });

        world.getEvents().getWorldStarted().on(null, true, () -> {
            if (movementTimer != null) {
                movementTimer.cancel();
                movementTimer = null;
            }
            if (animationTimer == null) {
                animationTimer = new Timer();
            }
            animationTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    world.getEvents().getFrameAdvanced().fireAndForget();
                }
            }, 50, 50);
        });

        world.getEvents().getWorldStopped().on(null, true, () -> {
            movementTimer = new Timer();
            movementTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    manualMovementUpdate();
                }
            }, 10, 10);
            if (animationTimer != null) {
                animationTimer.cancel();
                animationTimer = null;
            }
        });

        movementTimer = new Timer();
        movementTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                manualMovementUpdate();
            }
        }, 10, 10);

        world.getEvents().getTileMapChanged().fireAndForget();
    }

    private void renderAllLayers(OdorWorld world) {
        canvas.getLayer().removeAllChildren();
        layerImageList = world.getTileMap().createImageList();
        canvas.getLayer().addChildren(layerImageList);
        for (OdorWorldEntity oe : world.getEntityList()) {
            EntityNode node = new EntityNode(world, oe);
            canvas.getLayer().addChild(node);
        }
        repaint();
    }

    private void centerCameraToSelectedEntity() {
        if (!world.isUseCameraCentering()) {
            repaint();
            return;
        }

        if (!getSelectedModelEntities().isEmpty()) {
            double worldHeight = world.getHeight();
            double worldWidth = world.getWidth();
            PCamera camera = canvas.getCamera();
            PBounds cameraBounds = camera.getFullBounds();

            PNode firstNode = getFirstSelectedEntityNode();
            if (firstNode == null) {
                return;
            }

            PBounds firstNodeBounds = firstNode.getFullBounds();

            double cameraNewX = -cameraBounds.width / 2 + firstNodeBounds.x + firstNodeBounds.width / 2;
            double cameraNewY = -cameraBounds.height / 2 + firstNodeBounds.y + firstNodeBounds.height / 2;

            // Stop centering the entity if the view is going out of bound of the bound of the world.
            if (cameraNewX < 0) {
                cameraNewX = 0;
            }
            if (cameraNewY < 0) {
                cameraNewY = 0;
            }
            if (cameraNewX + cameraBounds.width > worldWidth) {
                cameraNewX = worldWidth - cameraBounds.width;
            }
            if (cameraNewY + cameraBounds.height > worldHeight) {
                cameraNewY = worldHeight - cameraBounds.height;
            }
            if (cameraBounds.width > worldWidth) {
                cameraNewX = 0;
            }
            if (cameraBounds.height > worldHeight) {
                cameraNewY = 0;
            }

            camera.setViewBounds(new Rectangle2D.Double(cameraNewX, cameraNewY, cameraBounds.width, cameraBounds.height));
            repaint();
        }
    }

    public List<Tile> getTileStack(Point2D point) {
        return world.getTileMap().getTileStackAtPixel(point);
    }

    public Tile getTile(Point2D point) {
        List<Tile> tileStack = world.getTileMap().getTileStackAtPixel(point);
        if (tileStack == null) {
            return null;
        } else {
            return tileStack.get(0);
        }
    }

    public void manualMovementUpdate() {
        EntityNode entityNode = getFirstSelectedEntityNode();
        if (entityNode != null && isManualMovementMode()) {
            OdorWorldEntity entity = entityNode.getEntity();
            entity.applyMovement();
            entityNode.advance();
            centerCameraToSelectedEntity();
        }
    }

    /**
     * Show the PNode debugging tool.
     */
    void showPNodeDebugger() {
        SceneGraphBrowser sgb = new SceneGraphBrowser(
            canvas.getRoot());
        StandardDialog dialog = new StandardDialog();
        dialog.setContentPane(sgb);
        dialog.setTitle("Piccolo Scenegraph Browser");
        dialog.setModal(false);
        dialog.pack();
        dialog.setLocationRelativeTo(null);
        dialog.setVisible(true);
    }


    /**
     * Get selected pnodes.
     */
    public List<PNode> getSelectedEntities() {
        // Assumes selected pnodes parents are entitynodes
        return getSelection().stream()
            .filter(p -> p.getParent() instanceof EntityNode)
            .map(p -> p.getParent())
            .collect(Collectors.toList());
    }

    /**
     * Get selected odor world entities.
     */
    public List<OdorWorldEntity> getSelectedModelEntities() {
        return getSelection().stream()
            .filter(p -> p.getParent() instanceof EntityNode)
            .map(p -> ((EntityNode) p.getParent()).getEntity())
            .collect(Collectors.toList());
    }

    public EntityNode getFirstSelectedEntityNode() {
        for (PNode n : getSelection()) {
            if (n.getParent() instanceof EntityNode) {
                return ((EntityNode) n.getParent());
            }
        }
        return null;
    }

    public OdorWorldEntity getFirstSelectedRotatingEntity() {
        return getSelectedModelEntities().stream().filter(e -> e.getEntityType().isRotating()).findFirst().orElse(null);
    }

    public OdorWorldEntity getFirstSelectedEntityModel() {
        if (getFirstSelectedEntityNode() != null) {
            return getFirstSelectedEntityNode().getEntity();
        }
        return null;
    }

    /**
     * Delete all current selected entities.
     */
    void deleteSelectedEntities() {
        for (OdorWorldEntity e : getSelectedModelEntities()) {
            e.delete();
        }
    }


    /**
     * Create a popup menu based on location of mouse click.
     *
     * @param entity the entity for which to build the menu
     * @return the popup menu
     */
    public JPopupMenu getContextMenu(OdorWorldEntity entity) {
        JPopupMenu contextMenu = new JPopupMenu();
        // No entity was clicked on
        if (entity == null) {
            contextMenu.add(new JMenuItem(new AddEntityAction(this)));
            contextMenu.add(new JMenuItem(new AddAgentAction(this)));
            contextMenu.add(new JMenuItem(new AddTile(this)));
            contextMenu.addSeparator();
        } else {
            contextMenu.add(new JMenuItem(new ShowEntityDialogAction(entity)));
            contextMenu.add(new JMenuItem(new DeleteEntityAction(this, entity)));
            contextMenu.addSeparator();

            // TODO: Create a delete smell source action
            if (entity.getSmellSource() == null) {
                contextMenu.add(new JMenuItem(new AddSmellSourceAction(this, entity)));
                contextMenu.addSeparator();
            }

            CouplingMenu couplingMenu = new CouplingMenu(component, entity);
            couplingMenu.setCustomName("Create couplings");
            contextMenu.add(couplingMenu);
            contextMenu.addSeparator();
        }
        contextMenu.add(new JMenuItem(new ShowWorldPrefsAction(this)));
        return contextMenu;
    }

    public OdorWorld getWorld() {
        return world;
    }

    public void clearSelection() {
        selectionModel.clear();
    }

    public Collection<PNode> getSelection() {
        return selectionModel.getSelection();
    }

    public PCanvas getCanvas() {
        return canvas;
    }

    public void setSelection(Collection elements) {
        selectionModel.setSelection(elements);
    }


    /**
     * Update selection handles.
     *
     * @param event the NetworkSelectionEvent
     */
    private void updateNodeHandles(final WorldSelectionEvent event) {

        Set<PNode> selection = event.getSelection();
        Set<PNode> oldSelection = event.getOldSelection();

        Set<PNode> difference = new HashSet<PNode>(oldSelection);
        difference.removeAll(selection);

        for (PNode node : difference) {
            NodeHandle.removeSelectionHandleFrom(node);
        }
        for (PNode node : selection) {
            // TODO: Move that to util class!
            NodeHandle.addSelectionHandleTo(node);

        }
    }

    /**
     * Return true if the specified element is selected.
     *
     * @param element element
     * @return true if the specified element is selected
     */
    public boolean isSelected(final Object element) {
        return selectionModel.isSelected(element);
    }


    /**
     * Toggle the selected state of the specified element; if it is selected,
     * remove it from the selection, if it is not selected, add it to the
     * selection.
     *
     * @param element element
     */
    public void toggleSelection(final Object element) {
        if (isSelected(element)) {
            selectionModel.remove(element);
        } else {
            selectionModel.add(element);
        }
    }

    private byte getManualMovementStateMask(String key) {
        byte mask = 0b0000;
        if ("w".equalsIgnoreCase(key)) {
            mask = 0b0001;
        } else if ("s".equalsIgnoreCase(key)) {
            mask = 0b0010;
        } else if ("a".equalsIgnoreCase(key)) {
            mask = 0b0100;
        } else if ("d".equalsIgnoreCase(key)) {
            mask = 0b1000;
        }
        return mask;
    }

    void setManualMovementKeyState(String key, boolean state) {
        byte mask = getManualMovementStateMask(key);
        if (state) {
            manualMovementKeyState |= mask;
        } else {
            manualMovementKeyState &= ~mask;
        }
    }

    boolean getManualMovementState(String key) {
        byte mask = getManualMovementStateMask(key);
        return (manualMovementKeyState & mask) > 0;
    }

    private boolean isManualMovementMode() {
        return manualMovementKeyState > 0;
    }

}

File: src/main/kotlin/org/simbrain/world/odorworld/OdorWorldPanel.kt
package org.simbrain.world.odorworld

import org.piccolo2d.PCanvas
import org.piccolo2d.PNode
import org.piccolo2d.event.PBasicInputEventHandler
import org.piccolo2d.event.PInputEvent
import org.piccolo2d.event.PInputEventListener
import org.piccolo2d.nodes.PImage
import org.piccolo2d.nodes.PPath
import org.piccolo2d.util.PPaintContext
import org.simbrain.network.gui.CustomToolBar
import org.simbrain.util.*
import org.simbrain.util.piccolo.*
import org.simbrain.world.odorworld.dialogs.EntityDialog
import org.simbrain.world.odorworld.entities.OdorWorldEntity
import org.simbrain.world.odorworld.gui.*
import java.awt.*
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.awt.event.MouseEvent
import java.awt.event.MouseWheelEvent
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D
import java.util.*
import java.util.Timer
import javax.swing.*
import kotlin.math.min
import kotlin.math.pow

/**
 * **OdorWorldPanel** represent the OdorWorld.
 */
class OdorWorldPanel(
    val odorWorldComponent: OdorWorldComponent,
    val world: OdorWorld
) : JPanel() {

    /**
     * The Piccolo PCanvas.
     */
    val canvas = OdorWorldCanvas()

    /**
     * Selection model.
     */
    val selectionManager = WorldSelectionManager()

    /**
     * Provisional interface for selecting tiles.
     */
    private var tileSelectionBox: PNode? = null
    private var tileSelectionModel: Rectangle? = null

    /**
     * Timer to update entity animations.
     */
    private var animationTimer: Timer? = null

    /**
     * Timer used for updating entity position when they are manually moved. Allows timing of
     * manualy movement to be consistent whether the workspace is running or not.
     */
    private var movementTimer: Timer? = null

    /**
     * Used for a mask that allows multiple movements to be applied at once.
     * Lower four bits are used for U,D,L,R.
     */
    private var manualMovementKeyState: Byte = 0

    /**
     * List corresponding to the layers of a tmx file.
     */
    private var layerImageList: List<PImage?>

    val odorWorldActions: OdorWorldActions = OdorWorldActions(this)

    /**
     * The current zoom level of the canvas.
     *
     * For example:
     * 0.5 means the canvas is rendered (zoomed out) at 0.5 of its normal size, and 2 means it is rendered (zoomed in) to twice its size
     *
     * The setter rescales the canvas.
     */
    var scalingFactor: Double
        get() = canvas.camera.viewScale
        set(scalingFactor) {
            val currentScalingFactor = canvas.camera.viewScale
            val scalingFactorRatio = scalingFactor / currentScalingFactor
            canvas.scale(scalingFactorRatio)
        }

    fun debugToolTips() {
        if (tileSelectionModel != null) {
            if (!tileSelectionModel!!.contains(world.lastClickedPosition)) {
                canvas.layer.removeChild(tileSelectionBox)
                tileSelectionModel = null
                tileSelectionBox = null
            }
        }

        if (tileSelectionBox == null) {
            val tileCoordinateX = (world.lastClickedPosition.x / world.tileMap.tileWidth).toInt()
            val tileCoordinateY = (world.lastClickedPosition.y / world.tileMap.tileHeight).toInt()
            tileSelectionModel = Rectangle(
                tileCoordinateX * world.tileMap.tileWidth,
                tileCoordinateY * world.tileMap.tileHeight,
                world.tileMap.tileWidth,
                world.tileMap.tileHeight
            )
            tileSelectionBox = PPath.createRectangle(
                tileSelectionModel!!.getX(),
                tileSelectionModel!!.getY(),
                tileSelectionModel!!.getWidth(),
                tileSelectionModel!!.getHeight()
            ).apply {
                strokePaint = Color.ORANGE
                setPaint(null)
            }
            canvas.layer.addChild(tileSelectionBox)
        }
    }

    /**
     * Extend PCanvas for custom handling of tooltips
     */
    inner class OdorWorldCanvas : PCanvas() {
        init {
            // Paradoxically, low quality rendering produces sharper quality for rendering pixel images
            // use nearest neighbor instead of bi-linear
            setDefaultRenderQuality(PPaintContext.LOW_QUALITY_RENDERING)
            animatingRenderQuality = PPaintContext.LOW_QUALITY_RENDERING
            interactingRenderQuality = PPaintContext.LOW_QUALITY_RENDERING
        }

        override fun getToolTipText(event: MouseEvent): String {
            val allNode = ArrayList((canvas.layer.allNodes as Collection<PNode?>).stream().toList())
            allNode.reverse()

            val firstNode = allNode.stream()
                .filter { it: PNode? ->
                    if (it is PImage) {
                        return@filter true
                    } else if (it is EntityNode) {
                        return@filter it.getFullBounds().contains(camera.localToView(event.point))
                    }
                    false
                }
                .findFirst()

            if (firstNode.isEmpty) {
                return ""
            }

            if (firstNode.get() is PImage) {
                val selectedTiles = getTileStack(event.point) ?: return ""
                val sb = StringBuilder("Tile Ids: ")
                selectedTiles.stream().filter { obj: Tile? -> Objects.nonNull(obj) }
                    .forEach { tile: Tile -> sb.append(" (" + tile.id + ")") }

                return sb.toString()
            } else if (firstNode.get() is EntityNode) {
                return (firstNode.get() as EntityNode).entity.toString()
            }
            return ""
        }

        /**
         * Change the current zoom level up or down by the scaling factor.
         *
         * For example:
         * 1.1 zooms in by ~10%
         * 0.9 zooms out by ~10%
         */
        fun scale(scalingFactor: Double) {
            val canvasCenter: Point2D = camera.bounds.center
            val (x, y) = camera.localToView(canvasCenter)
            val newWidth = camera.viewBounds.width / scalingFactor
            val newHeight = camera.viewBounds.height / scalingFactor
            val newX = x - newWidth / 2
            val newY = y - newHeight / 2
            canvas.setViewBounds(Rectangle2D.Double(newX, newY, newWidth, newHeight))
        }

        fun setViewBounds(bounds: Rectangle2D) {
            val (x, y, w, h) = bounds
            val newWidth = min(w, world.width)
            val newHeight = min(h, world.height)
            val newX = x.coerceIn(0.0, world.width - newWidth)
            val newY = y.coerceIn(0.0, world.height - newHeight)
            canvas.camera.setViewBoundsNoOverflow(
                Rectangle2D.Double(
                    newX,
                    newY,
                    newWidth,
                    newHeight
                )
            )
        }
    }

    /**
     * Construct a world, set its background color.
     *
     * @param world the frame in which this world is rendered
     */
    init {
        ToolTipManager.sharedInstance().registerComponent(canvas)

        layout = BorderLayout()
        this.add("Center", canvas)
        add("North", createMainToolBar())

        canvas.isFocusable = true

        // Add tile map
        layerImageList = world.tileMap.createImageList()
        canvas.layer.addChildren(layerImageList)

        // Remove default event handlers
        val panEventHandler: PInputEventListener = canvas.panEventHandler
        val zoomEventHandler: PInputEventListener = canvas.zoomEventHandler
        canvas.removeInputEventListener(panEventHandler)
        canvas.removeInputEventListener(zoomEventHandler)

        // Add key bindings
        OdorWorldKeyBindings.addBindings(this)

        // Mouse events
        canvas.addInputEventListener(WorldMouseHandler(this, world))
        canvas.addInputEventListener(WorldContextMenuEventHandler(this, world))

        world.events.entityAdded.on(swingDispatcher) { e: OdorWorldEntity? ->
            val node = EntityNode(e!!)
            canvas.layer.addChild(node)
            selectionManager.clear()
            selectionManager.add(node)
        }
        world.events.updated.on(swingDispatcher, wait = true) { this.centerCameraToSelectedEntity() }
        world.events.frameAdvanced.on(swingDispatcher) {
            canvas.layer.childrenReference
                .filterIsInstance<EntityNode>()
                .forEach { it.advance() }
            repaint()
        }
        world.events.animationStopped.on(swingDispatcher) {
            // When movement is stopped use the "static" animation so we don't show entities in strange
            // intermediate states
            canvas.layer.childrenReference
                .filterIsInstance<EntityNode>()
                .forEach { it.resetToStaticFrame() }
            repaint()
        }

        // Full tile map update
        world.events.tileMapChanged.on(swingDispatcher) {
            world.tileMap.events.layersChanged.on(swingDispatcher) {
                renderAllLayers(world)
            }
            world.tileMap.events.mapSizeChanged.on(swingDispatcher) {
                world.events.tileMapChanged.fire()
            }

            // Single layer update
            world.tileMap.events.layerImageChanged.on(
                wait = true, dispatcher = swingDispatcher
            ) { oldImage: PImage?, newImage: PImage? ->
                val index = canvas.layer.indexOfChild(oldImage)
                canvas.layer.removeChild(oldImage)
                if (index != -1) {
                    canvas.layer.addChild(index, newImage)
                } else {
                    canvas.layer.addChild(newImage)
                }
            }
            renderAllLayers(world)
        }

        world.events.worldStarted.on(null, true) {
            if (movementTimer != null) {
                movementTimer?.cancel()
                movementTimer = null
            }
            if (animationTimer == null) {
                animationTimer = Timer().apply {
                    schedule(object : TimerTask() {
                        override fun run() {
                            world.events.frameAdvanced.fire()
                        }
                    }, 50, 50)
                }
            }
        }

        world.events.worldStopped.on(null, true, Runnable {
            movementTimer = Timer().apply {
                schedule(object : TimerTask() {
                    override fun run() {
                        manualMovementUpdate()
                    }
                }, 10, 10)
            }
            if (animationTimer != null) {
                animationTimer?.cancel()
                animationTimer = null
            }
        })

        canvas.addInputEventListener(object : PBasicInputEventHandler() {
            override fun mouseWheelRotated(event: PInputEvent) {
                val swingEvent = (event.sourceSwingEvent as MouseWheelEvent)
                val newScale = 1.1.pow(swingEvent.preciseWheelRotation)
                canvas.scale(1 / newScale)
            }
        })

        movementTimer = Timer().apply {
            schedule(object : TimerTask() {
                override fun run() {
                    manualMovementUpdate()
                }
            }, 10, 10)
        }

        world.events.tileMapChanged.fire()

        canvas.setViewBounds(Rectangle2D.Double(0.0, 0.0, world.width, world.height))

        // Repaint whenever window is opened or changed.
        addComponentListener(object : ComponentAdapter() {
            override fun componentResized(arg0: ComponentEvent) {
                scalingFactor = scalingFactor // force invoke setter
            }
        })

        odorWorldActions.createSelectAllAction()
    }

    private fun renderAllLayers(world: OdorWorld) {
        canvas.layer.removeAllChildren()
        layerImageList = world.tileMap.createImageList()
        canvas.layer.addChildren(layerImageList)
        for (oe in world.entityList) {
            val node = EntityNode(oe)
            canvas.layer.addChild(node)
        }
        repaint()
    }

    private fun centerCameraToSelectedEntity() {
        if (!world.isUseCameraCentering) {
            repaint()
            return
        }

        if (selectedEntityModels.isNotEmpty()) {
            canvas.setViewBounds(
                Rectangle2D.Double(
                    firstSelectedEntityModel!!.x - canvas.camera.viewBounds.width / 2,
                    firstSelectedEntityModel!!.y - canvas.camera.viewBounds.height / 2,
                    canvas.camera.viewBounds.width,
                    canvas.camera.viewBounds.height
                )
            )
            repaint()
        }
    }

    fun getTileStack(point: Point2D?): List<Tile> {
        return world.tileMap.getTileStackAtPixel(point!!)
    }

    fun getTile(point: Point2D?): Tile? {
        val tileStack = world.tileMap.getTileStackAtPixel(
            point!!
        )
        return if (tileStack == null) {
            null
        } else {
            tileStack[0]
        }
    }

    fun manualMovementUpdate() {
        val entityNode = firstSelectedEntityNode
        if (entityNode != null && isManualMovementMode) {
            val entity = entityNode.entity
            entity.applyMovement()
            entityNode.advance()
            centerCameraToSelectedEntity()
        }
    }

    /**
     * Show the PNode debugging tool.
     */
    fun showPNodeDebugger() {
        val sgb = SceneGraphBrowser(
            canvas.root
        )
        val dialog = StandardDialog()
        dialog.contentPane = sgb
        dialog.title = "Piccolo Scenegraph Browser"
        dialog.isModal = false
        dialog.pack()
        dialog.setLocationRelativeTo(null)
        dialog.isVisible = true
    }


    val selectedEntityNodes: List<EntityNode>
        get() =
            selection
                .map {
                    if (it is EntityNode) {
                        it
                    } else {
                        it.parent
                    }
                }
                .filterIsInstance<EntityNode>()

    val selectedEntityModels: List<OdorWorldEntity>
        get() = selectedEntityNodes
            .map { it.entity }

    val firstSelectedEntityNode: EntityNode?
        get() = selectedEntityNodes.firstOrNull()

    val firstSelectedRotatingEntity: OdorWorldEntity?
        get() = selectedEntityModels.firstOrNull { it.entityType.isRotating }

    val firstSelectedEntityModel: OdorWorldEntity?
        get() {
            if (firstSelectedEntityNode != null) {
                return firstSelectedEntityNode!!.entity
            }
            return null
        }

    /**
     * Delete all current selected entities.
     */
    fun deleteSelectedEntities() {
        selectedEntityModels.forEach { it.delete() }
    }

    /**
     * Popup menu when not clicking on entity. On an entity see [EntityNode.createContextMenu]
     */
    fun getContextMenu() = JPopupMenu().apply {
        add(JMenuItem(odorWorldActions.addAgentAction()))
        add(JMenuItem(odorWorldActions.addEntityAction()))
        addSeparator()
        add(JMenuItem(odorWorldActions.addTileAction))
        add(JMenuItem(odorWorldActions.fillLayerAction))
        add(odorWorldActions.createChooseLayerMenu(world))
        add(odorWorldActions.editLayersAction)
        addSeparator()
        add(JMenuItem(odorWorldActions.showWorldPrefsAction()))
    }

    fun clearSelection() {
        selectionManager.clear()
    }

    var selection: MutableSet<PNode>
        get() = selectionManager.selection
        set(elements) {
            selectionManager.clear()
            selectionManager.addAll(elements)
        }

    /**
     * Return true if the specified element is selected.
     *
     * @param element element
     * @return true if the specified element is selected
     */
    fun isSelected(element: Any?): Boolean {
        return selectionManager.isSelected(element)
    }


    /**
     * Toggle the selected state of the specified element; if it is selected,
     * remove it from the selection, if it is not selected, add it to the
     * selection.
     *
     * @param element element
     */
    fun toggleSelection(element: PNode) {
        if (isSelected(element)) {
            selectionManager.remove(element)
        } else {
            selectionManager.add(element)
        }
    }

    private fun getManualMovementStateMask(key: String): Byte {
        var mask: Byte = 0
        if ("w".equals(key, ignoreCase = true)) {
            mask = 1
        } else if ("s".equals(key, ignoreCase = true)) {
            mask = 2
        } else if ("a".equals(key, ignoreCase = true)) {
            mask = 4
        } else if ("d".equals(key, ignoreCase = true)) {
            mask = 8
        }
        return mask
    }

    fun setManualMovementKeyState(key: String, state: Boolean) {
        val mask = getManualMovementStateMask(key)
        manualMovementKeyState = if (state) {
            (manualMovementKeyState.toInt() or mask.toInt()).toByte()
        } else {
            (manualMovementKeyState.toInt() and mask.toInt().inv()).toByte()
        }
    }

    fun getManualMovementState(key: String): Boolean {
        val mask = getManualMovementStateMask(key)
        return (manualMovementKeyState.toInt() and mask.toInt()) > 0
    }

    private val isManualMovementMode: Boolean
        get() = manualMovementKeyState > 0

    private fun createMainToolBar() = CustomToolBar().apply {
        add(odorWorldActions.addAgentAction())
        add(odorWorldActions.addEntityAction())
        addSeparator()
        add(odorWorldActions.resetZoomAction())
        add(odorWorldActions.zoomInAction())
        add(odorWorldActions.zoomOutAction())
    }

    fun editSelectedEntities() {
        selectedEntityNodes.firstOrNull()?.let {
            EntityDialog(it.entity).apply { title = "Edit ${it.entity.name}" }.display()
        }
    }
}

